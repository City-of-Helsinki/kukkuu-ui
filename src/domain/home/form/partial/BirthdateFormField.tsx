import React, { FunctionComponent } from 'react';
import { FieldArrayRenderProps, getIn } from 'formik';
import { useTranslation } from 'react-i18next';

import styles from './birthdateFormField.module.scss';
import { validateRequire } from '../../../../common/components/form/validationUtils';
import EnhancedInputField from '../../../../common/components/form/fields/input/EnhancedInputField';
import NumberInputField from '../../../../common/components/form/fields/input/NumberInputField';
import {
  formatTime,
  newMoment,
  toZeroBasedMonth,
} from '../../../../common/time/utils';
import { DEFAULT_DATE_FORMAT } from '../../../../common/time/TimeConstants';
import { Birthdate } from '../../../child/form/ChildForm';

interface BirthdateFormFieldProps extends FieldArrayRenderProps {
  isImmutable?: boolean;
  values?: Birthdate;
}

const BirthdateFormField: FunctionComponent<BirthdateFormFieldProps> = ({
  name,
  isImmutable = false,
  values,
  form: { errors, touched },
}) => {
  const fieldTouched =
    getIn(touched, 'child.birthdate.day') || getIn(touched, 'birthdate.day');
  const error = getIn(errors, 'childBirthdate');
  const { t } = useTranslation();

  if (isImmutable && values) {
    const monthZeroBased = toZeroBasedMonth(values.month);
    const birthdate = formatTime(
      newMoment([values.year, monthZeroBased, values.day]),
      DEFAULT_DATE_FORMAT
    );

    return (
      <div className={styles.birthdateField}>
        <label>{`${t(
          'homePage.preliminaryForm.childBirthdate.input.label'
        )}*`}</label>
        <p className={styles.immutableField}>{birthdate}</p>
      </div>
    );
  }

  return (
    <div className={styles.birthdateField}>
      <label>{`${t(
        'homePage.preliminaryForm.childBirthdate.input.label'
      )}*`}</label>
      <div className={styles.inputWrapper}>
        <EnhancedInputField
          name={`${name}.day`}
          component={NumberInputField}
          aria-label={t(
            'homePage.preliminaryForm.childBirthdate.input.day.placeholder'
          )}
          placeholder={t(
            'homePage.preliminaryForm.childBirthdate.input.day.placeholder'
          )}
          required={true}
          min={1}
          max={31}
          maxLength={2}
        />
        <div className={styles.dot}>.</div>
        <EnhancedInputField
          name={`${name}.month`}
          required={true}
          component={NumberInputField}
          aria-label={t(
            'homePage.preliminaryForm.childBirthdate.input.month.placeholder'
          )}
          placeholder={t(
            'homePage.preliminaryForm.childBirthdate.input.month.placeholder'
          )}
          min={1}
          max={12}
          maxLength={2}
        />
        <div className={styles.dot}>.</div>
        <EnhancedInputField
          required={true}
          name={`${name}.year`}
          maxLength={4}
          component={NumberInputField}
          validate={(value: number) => validateRequire(value)}
          aria-label={t(
            'homePage.preliminaryForm.childBirthdate.input.year.placeholder'
          )}
          placeholder={t(
            'homePage.preliminaryForm.childBirthdate.input.year.placeholder'
          )}
        />
      </div>
      {/* not to display error at first render until input got touched */}
      {fieldTouched && <div className={styles.error}>{t(error)}</div>}
    </div>
  );
};

export default BirthdateFormField;
