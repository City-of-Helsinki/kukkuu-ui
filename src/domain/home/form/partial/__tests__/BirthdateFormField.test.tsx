import { FieldArray } from 'formik';

import BirthdateFormField from '../BirthdateFormField';
import TestForm from '../../../../../common/test/TestForm';
import { render } from '../../../../../common/test/testingLibraryUtils';

it('renders snapshot correctly', () => {
  const { container } = render(
    <TestForm>
      <FieldArray
        name="foo"
        render={(props) => <BirthdateFormField {...props} />}
      />
    </TestForm>
  );
  expect(container).toMatchSnapshot();
});
