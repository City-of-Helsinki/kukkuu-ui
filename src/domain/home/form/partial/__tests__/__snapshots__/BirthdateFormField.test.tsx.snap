// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot correctly 1`] = `"<div class=\\"birthdateField\\"><label>Lapsen syntym√§aika (pp.kk.vvvv)*</label><div class=\\"inputWrapper\\"><div class=\\"TextInput-module_root__1ceal text-input_hds-text-input__24bl6\\"><div class=\\"TextInput-module_inputWrapper__1ru3h text-input_hds-text-input__input-wrapper__2p7Uy\\"><input type=\\"number\\" class=\\"TextInput-module_input__35GDy text-input_hds-text-input__input__2XAyI\\" id=\\"foo.day\\" required=\\"\\" name=\\"foo.day\\" aria-label=\\"pv\\" placeholder=\\"pv\\" min=\\"1\\" max=\\"31\\" maxLength=\\"2\\"/></div></div><div class=\\"dot\\">.</div><div class=\\"TextInput-module_root__1ceal text-input_hds-text-input__24bl6\\"><div class=\\"TextInput-module_inputWrapper__1ru3h text-input_hds-text-input__input-wrapper__2p7Uy\\"><input type=\\"number\\" class=\\"TextInput-module_input__35GDy text-input_hds-text-input__input__2XAyI\\" required=\\"\\" name=\\"foo.month\\" aria-label=\\"kk\\" placeholder=\\"kk\\" min=\\"1\\" max=\\"12\\" maxLength=\\"2\\"/></div><div class=\\"TextInput-module_helperText__anJHZ text-input_hds-text-input__helper-text__X1cCW\\"></div></div><div class=\\"dot\\">.</div><div class=\\"TextInput-module_root__1ceal text-input_hds-text-input__24bl6\\"><div class=\\"TextInput-module_inputWrapper__1ru3h text-input_hds-text-input__input-wrapper__2p7Uy\\"><input type=\\"number\\" class=\\"TextInput-module_input__35GDy text-input_hds-text-input__input__2XAyI\\" required=\\"\\" name=\\"foo.year\\" maxLength=\\"4\\" min=\\"2019\\" aria-label=\\"vuosi\\" placeholder=\\"vuosi\\"/></div></div></div></div>"`;
